imbi/endpoints/project_ssm_configuration.py

```diff
@@ -0,0 +1,216 @@
+        cached_creds = await self._redis.hgetall(key)
+        if cached_creds:
+            return {
+                key.decode('utf-8'): value.decode('utf-8')
+                for key, value in cached_creds.items()
+            }
+        try:
+            response = await aws.get_credentials(aws_session, role_arn,
+                                                 imbi_user.username, id_token)
+        except botocore.exceptions.ClientError as error:
+            if not is_retry and error.response['Error'][
+                    'Code'] == 'ExpiredTokenException':
+                id_token = await self.refresh_oauth2_tokens(
+                    imbi_user.username, imbi_user.external_id, refresh_token)
+                return await self.get_aws_credentials(aws_session, role_arn,
+                                                      imbi_user, id_token,
+                                                      refresh_token, True)
+            else:
+                error_code = error.response['Error']['Code']
+                if error_code == 'AccessDenied':
+                    raise web.HTTPError(
+                        401,
+                        reason=('Not authorized to access SSM in this '
+                                'namespace & environment'))
+                else:
+                    self.logger.error('Unexpected AWS credential failure: %s',
+                                      error_code)
+                    raise web.HTTPError(
+                        500,
+                        reason='Unexpected failure fetching AWS credentials')
+        creds = {
+            'access_key_id': response['Credentials']['AccessKeyId'],
+            'secret_access_key': response['Credentials']['SecretAccessKey'],
+            'session_token': response['Credentials']['SessionToken']
+        }
+        await self._redis.hmset_dict(key, creds)
+        await self._redis.expireat(
+            key, response['Credentials']['Expiration'].timestamp())
+        return creds
+
+    async def refresh_oauth2_tokens(self, username, external_id,
+                                    refresh_token) -> str:
+        env = os.environ.get('ENVIRONMENT', 'production').lower()
+        google_integration = await oauth2.OAuth2Integration.by_name(
+            self.application, f'google-{env}')
+        body = urllib.parse.urlencode({
+            'grant_type': 'refresh_token',
+            'redirect_uri': google_integration.callback_url,
+            'refresh_token': refresh_token,
+        })
+        response = await self.http_fetch(
+            str(google_integration.token_endpoint),
+            method='POST',
+            body=body,
+            content_type='application/x-www-form-urlencoded',
+            auth_username=google_integration.client_id,
+            auth_password=google_integration.client_secret)
+        if response.code in (401, 403):
+            await self.session.clear()
+            raise web.Finish()
```

Comment:
ek, last I tested this it just logged you out, which is what I wanted. Now you get a 200 response, no data for the table, and nothing happens till you try to navigate somewhere, and then you're taken to the login screen (since you no longer have a valid session).

I was only considering the case where the token was revoked/expired. If we plan to narrow that down, I guess I need to test this by prepping a log line and waiting for it to be revoked. Google really doesn't have proper docs on the responses to these endpoints.

Reply:
I did end up returning a 403. Had to figure out how to handle the redirect on the frontend instead of the backend, since this is an async request. The corresponding FE commit for this is: https://github.com/AWeber-Imbi/imbi-ui/pull/72/commits/fb5bf32e24d5020358b90639018b36c05ef9c54c